core/src/main/java/io/aiven/klaw/service/CacheService.java

```diff
@@ -0,0 +1,158 @@
+package io.aiven.klaw.service;
+
+import java.util.*;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.core.ParameterizedTypeReference;
+import org.springframework.http.*;
+import org.springframework.web.client.RestClientException;
+import org.springframework.web.client.RestTemplate;
+
+@Slf4j
+public class CacheService<T> {
+
+  public static final String URL_SEPERATOR = "/";
+  public static final String TENANT = "tenant";
+  public static final String ID = "id";
+  Map<Integer, Map<Integer, T>> cache;
+
+  RestTemplate rest;
+
+  private String urlEndpoint;
+
+  private HighAvailabilityUtilsService utilsService;
+
+  private String clusterUrlsAsString;
+
+  private List<String> clusterUrls;
+
+  public CacheService(
+      String endpoint, String clusterUrlsAsString, HighAvailabilityUtilsService utilsService) {
+    // TODO an interface needs to be added to allow the passing in of different communication
+    // methods (kafka/https/rabbitmq etc) to allow any org to use what they want for maintaining
+    // cache.
+    this.cache = new HashMap<>();
+    this.urlEndpoint = "cache" + URL_SEPERATOR + endpoint;
+    this.clusterUrlsAsString = clusterUrlsAsString;
+    this.utilsService = utilsService;
+  }
+
+  public T add(int tenantId, Integer id, T entry) {
+    return addOrUpdate(tenantId, id, entry, false);
+  }
+
+  public void addAll(int tenantId, Map<Integer, T> entries) {
+    if (entries == null) {
+      entries = new HashMap<>();
+    }
+    getCache(tenantId).putAll(entries);
+  }
+
+  public T remove(int tenantId, Integer id, boolean isLocalUpdate) {
```

Comment:
is this id always environment id ?

Reply:
No it can be the teamId, TopicId, SchemaId anything that we might cache, this Cache can be used to add a common HA and access/removal etc method for all information we want to cache, without having to reload the DB on any other instances in the cluster.

Reply:
How to identify the id type ?